### PR TITLE
Make the usage of Guava optional in all modules

### PR DESCRIPTION
**Pull request by joschi**
Refs #2535

### CODE REVIEW

File: dropwizard-jersey/src/main/java/io/dropwizard/jersey/DropwizardResourceConfig.java [hunk 1]
```diff
@@ -67,17 +67,20 @@ public DropwizardResourceConfig(@Nullable MetricRegistry metricRegistry) {
...
-        register(io.dropwizard.jersey.guava.OptionalMessageBodyWriter.class);
-        register(new io.dropwizard.jersey.guava.OptionalParamBinder());
-        register(io.dropwizard.jersey.optional.OptionalMessageBodyWriter.class);
-        register(io.dropwizard.jersey.optional.OptionalDoubleMessageBodyWriter.class);
-        register(io.dropwizard.jersey.optional.OptionalIntMessageBodyWriter.class);
-        register(io.dropwizard.jersey.optional.OptionalLongMessageBodyWriter.class);
+        register(new CacheControlledResponseFeature());
+        register(new io.dropwizard.jersey.guava.OptionalMessageBodyWriter());
+        register(new io.dropwizard.jersey.optional.OptionalMessageBodyWriter());
+        register(new io.dropwizard.jersey.optional.OptionalDoubleMessageBodyWriter());
```
Comment (jplock) [inline dropwizard-jersey/src/main/java/io/dropwizard/jersey/DropwizardResourceConfig.java]: Should we register these as instances or classes going forward? I thought part of the issue with the latest Jersey was that it no longer supported registering instances?

**Comment by mattnelson** [comment]
> Copy RateLimiter (and dependent classes) from Guava 27.0

I think a shade with [include](https://maven.apache.org/plugins/maven-shade-plugin/examples/includes-excludes.html) and [relocate](https://maven.apache.org/plugins/maven-shade-plugin/examples/class-relocation.html) would be a more preferable option to strait copies in order to continue picking up upstream changes to guava.

**Comment by Tibor17** [comment]
Yes, I mentioned `maven-shade-plugin` too. We use it in Apache Maven but you have to check the license of guava if you want to inline it in dropwizard.

**Comment by minisu** [comment]
Did you consider moving the Guava stuff to a `dropwizard-guava` module instead (similar to how Java 8 specifics used to reside in [dropwizard-java8](https://github.com/dropwizard/dropwizard-java8))?

**Comment by Tibor17** [comment]
There are still 8 open issues in version `2.0`.
Can the Dropwizard committers make a decision about this issue and make a bigger progress on it?

**Comment by joschi** [comment]
@dropwizard/committers Given that many people are still using Guava when building Dropwizard applications and Guava's improved focus on backward-compatibility, I'm not sure if pursuing the removal in Dropwizard 2.x is still worth it.

Any opinions?

**Comment by jplock** [comment]
@joschi makes sense to me. I think we can leave Guava for now.

**Comment by nickbabcock** [comment]
Right now we're kinda in a halfway state, where we've incorporated a subset of Guava classes (like `Lists`). I'm fine with continuing with guava or removing it -- but a halfway state doesn't seem like the best long term solution

**Comment by jplock** [comment]
I think of there are jdk8 compatible replacements, we should opt for using those whenever possible. For anything else using guava is ok.

**Comment by Tibor17** [comment]
Guava is breaking backwards compatibility. It is found as a problematic library nevertheless historically many people have used it before. One way would be to remove it completely or write utilities in Dropwizard which substitutes Guava and the last alternative would be to use `maven-shade-plugin` and shift Java packages and inline into dropwizard libs.

**Comment by joschi** [comment]
@Tibor17 Every library is eventually breaking backward compatibility. Pretty much every single dependency of Dropwizard did backward-incompatible changes in the past, so that's not a really good argument in my opinion.

Sure, we could shade it, but following that precedent, we'd also have to shade other libraries. 🤷‍♂️

**Comment by Tibor17** [comment]
@joschi 
Removing Guava happens elsewhere but here important is to make stable Dropwizards towards users since it is a big project. As you can see Google breaks typical rules of what major.minor.incremental version means and you can really expect broken compatibility with increased minor version. Next bad with Guava is the modularity because this is libbrary having multiple responsibilities instead of a good design with modular Guava which was in past.

**Pull request closed without merging**